fsharp-backend/src/LibExecution/Interpreter.fs

```diff
@@ -237,164 +238,94 @@ let rec eval' (state : ExecutionState) (st : Symtable) (e : Expr) : DvalTask =
+        | PBool (id, b) -> (dv = DBool b), [], [ (id, DBool b) ]
+        | PCharacter (id, c) -> (dv = DChar c), [], [ (id, DChar c) ]
+        | PString (id, s) -> (dv = DStr s), [], [ (id, DStr s) ]
+        | PFloat (id, f) -> (dv = DFloat f), [], [ (id, DFloat f) ]
+        | PNull (id) -> (dv = DNull), [], [ (id, DNull) ]
+
+        | PBlank (id) -> false, [], [ (id, incomplete id) ]
+
+        | PVariable (id, varName) ->
+          not (Dval.isFake dv), [ (varName, dv) ], [ (id, dv) ]
+
+        | PConstructor (id, name, args) ->
+          let traceArgsIncomplete argPatterns =
+            argPatterns
+            |> List.map Pattern.toID
+            |> List.map (fun pId -> (pId, incomplete pId))
+
+          match (name, args, dv) with
+          | "Nothing", [], v -> (v = DOption None), [], [ (id, DOption None) ]
+
+          | "Just", [ p ], DOption (Some v)
+          | "Ok", [ p ], DResult (Ok v)
+          | "Error", [ p ], DResult (Error v) ->
+            let (passes, newVars, traces) = checkPattern v p
+
+            if passes then
+              true, newVars, [ (id, dv) ] @ traces
+            else
+              false,
+              newVars,
+              (id, incomplete id) :: traceArgsIncomplete [ p ] @ traces
+
+          | _name, argPatterns, _dv ->
+            let traces = (id, incomplete id) :: traceArgsIncomplete argPatterns
+            false, [], traces
+
+      // This is to avoid checking `state.tracing.realOrPreview = Real` below.
+      // If RealOrPreview gets additional branches, reconsider what to do here.
+      let isRealExecution =
+        match state.tracing.realOrPreview with
+        | Real -> true
+        | Preview -> false
+
+      // The value we're matching against
       let! matchVal = eval state st matchExpr
 
-      do!
-        Ply.List.iterSequentially
-          (fun (pattern, expr) -> matchAndExecute matchVal [] (pattern, expr))
-          cases
+      let mutable hasMatched = false
+      let mutable matchResult = incomplete id
+
+      for (pattern, rhsExpr) in cases do
+        if hasMatched && isRealExecution then
+          ()
+        else
+          let passes, newDefs, traces = checkPattern matchVal pattern
+          let newSymtable = Map.mergeFavoringRight st (Map.ofList newDefs)
+
+          if not hasMatched && passes then
+            traces
+            |> List.iter (fun (id, dv) ->
+              state.tracing.traceDval state.onExecutionPath id dv)
+
+            let! r = eval state newSymtable rhsExpr
+            matchResult <- r
+            hasMatched <- true
+          else if isRealExecution then
+            // "Real" evaluations don't need to persist non-matched traces
+            ()
+          else
+            // If we're "previewing" (analysis), persist traces for all patterns
+            traces |> List.iter (fun (id, dv) -> state.tracing.traceDval false id dv)
+            state.tracing.traceDval false id matchVal
```

Comment:
I'm confused by this. Why is it required on this branch (or conversely, why is it not required on the other branch)?

Reply:
Hmm, I forget! I'll have to check.

The fuzztesting caught this - I think this ends up _overwriting_ the trace in some cases? I'll research in a bit and follow up.

Reply:
I just went back to a commit where the fuzztests were still valid (before removing old code, etc.) to test this.

I removed that line, ran fuzztests for a while, and couldn't reproduce whatever I was seeing before, and otherwise forget the details.
I suppose I'll remove it now - if something comes up where we figure this out, great.